presto-main/src/main/java/com/facebook/presto/operator/aggregation/AbstractGroupCollectionAggregationState.java

```diff
@@ -0,0 +1,185 @@
+    private static final int MAX_NUM_BLOCKS = 30000;
+    private static final short NULL = -1;
+
+    private final ShortBigArray headBlockIndex;
+    private final IntBigArray headPosition;
+
+    private final ShortBigArray nextBlockIndex;
+    private final IntBigArray nextPosition;
+
+    private final ShortBigArray tailBlockIndex;
+    private final IntBigArray tailPosition;
+
+    private final List<PageBuilder> values;
+    private final LongList sumPositions;
+    private final IntBigArray groupEntryCount;
+    private PageBuilder currentPageBuilder;
+
+    private long valueBlocksRetainedSizeInBytes;
+    private long totalPositions;
+    private long capacity;
+
+    protected AbstractGroupCollectionAggregationState(PageBuilder pageBuilder)
+    {
+        this.headBlockIndex = new ShortBigArray(NULL);
+        this.headPosition = new IntBigArray(NULL);
+        this.nextBlockIndex = new ShortBigArray(NULL);
+        this.nextPosition = new IntBigArray(NULL);
+        this.tailBlockIndex = new ShortBigArray(NULL);
+        this.tailPosition = new IntBigArray(NULL);
+
+        this.currentPageBuilder = pageBuilder;
+        this.values = new ArrayList<>();
+        this.sumPositions = new LongArrayList();
+        this.groupEntryCount = new IntBigArray();
+        values.add(currentPageBuilder);
+        sumPositions.add(0L);
+        valueBlocksRetainedSizeInBytes = 0;
+
+        totalPositions = 0;
+        capacity = 1024;
+        nextBlockIndex.ensureCapacity(capacity);
+        nextPosition.ensureCapacity(capacity);
+        groupEntryCount.ensureCapacity(capacity);
+    }
+
+    @Override
+    public void ensureCapacity(long size)
+    {
+        headBlockIndex.ensureCapacity(size);
+        headPosition.ensureCapacity(size);
+        tailBlockIndex.ensureCapacity(size);
+        tailPosition.ensureCapacity(size);
+        groupEntryCount.ensureCapacity(size);
+    }
+
+    @Override
+    public long getEstimatedSize()
+    {
+        return INSTANCE_SIZE +
+                headBlockIndex.sizeOf() +
+                headPosition.sizeOf() +
+                tailBlockIndex.sizeOf() +
+                tailPosition.sizeOf() +
+                nextBlockIndex.sizeOf() +
+                nextPosition.sizeOf() +
+                groupEntryCount.sizeOf() +
+                valueBlocksRetainedSizeInBytes +
+                // valueBlocksRetainedSizeInBytes doesn't contain the current block builder
+                currentPageBuilder.getRetainedSizeInBytes();
+    }
+
+    /**
+     * Performs the necessary recordkeeping of internal linked list to prepare for a new entry in the underlying PageBuilder.
+     * This method should be called before {@link #appendAtChannel(int, Block, int)} to update the internal linked list, where
+     * {@link #appendAtChannel(int, Block, int)} is called for each channel that has a new entry to be added.
+     */
+    protected final void prepareAdd()
+    {
+        if (currentPageBuilder.isFull()) {
```

Comment:
I did that because it didn't seem right to shift position recordkeeping inside the current page builder, and then also do the same position counting here, when I can rely on just the former.  It seemed safe since the scope of usage of this variable is purely inside this method.

Reply:
Per offline discussion, I reverted to old `totalPositions` semantics.